client/gutenberg/editor/controller.js

```diff
@@ -38,22 +40,50 @@ function getPostID( context ) {
 	return parseInt( context.params.post, 10 );
 }
 
-export const redirect = ( context, next ) => {
+function waitForSiteIdAndSelectedEditor( context ) {
+	return new Promise( resolve => {
+		const unsubscribe = context.store.subscribe( () => {
+			const state = context.store.getState();
+			const siteId = getSelectedSiteId( state );
+			if ( ! siteId ) {
+				return;
+			}
+			const selectedEditor = getSelectedEditor( state, siteId );
+			if ( ! selectedEditor ) {
+				return;
+			}
+			unsubscribe();
+			resolve();
+		} );
+		// Trigger a `store.subscribe()` callback
+		context.store.dispatch(
+			requestSelectedEditor( getSelectedSiteId( context.store.getState() ) )
+		);
+	} );
```

Comment:
if we refactor to use `http-data` here we should be able to further simplify this into `waitForData()`

```js
import { empty as emptyData, waitForData } from 'state/http-data';

const { selectedEditor } = await waitForData( {
	selectedEditor: () => {
		const siteId = getSelectedSiteId( context.store.getState() );
		return siteId
			? requestSelectedEditor( siteId )
			: emptyData;
	}
} );
```

Reply:
actually we would move the `siteId` check _into_ `requestSelectedEditor()`

```js
const { selectedEditor } = await waitForData( {
	selectedEditor: () => requestSelectedEditor( context.store.getState() ),
} );

export const requestSelectedEditor = siteId => {
	if ( ! siteId ) {
		return empty;
	}

	const id = `selected-editor-${ siteId }`;

	return requestHttpData(
		id,
		http( … ),
		{
			fromApi: () => ( { editor } ) => {
				switch ( editor ) {
					case 'classic':
					case 'gutenberg-iframe':
					case 'gutenberg-redirect':
					case 'gutenberg-redirect-and-style':
						return [ [ id, editor ] ];

					default:
						throw new Error( `unknown editor for site: ${ editor }` );
				}
			}
		}
	);
}
```

Reply:
Thanks for the suggestion @dmsnell!

It's actually a great idea, but after trying to use this approach, I think this requires a bigger refactor, so we're going to defer it for a later time.

The main problem is that we don't only need to wait for the data to be fetched, we also have to update the state so the `getSelectedEditor` selector returns the expected value.

We could of course update the state after waiting for the data to be fetched, but this is barely changing the behavior we had before (the state was automatically updated when the action was dispatched). 

An alternative would be to change the selector to retrieve the selected editor value using `getHttpData`, but this will require additional changes to the action we dispatch for setting a new selected editor.

Given that the above are not trivial changes, I consider ok to leave things as they are.